packages/flutter_tools/lib/src/mdns_discovery.dart

```diff
@@ -229,10 +230,28 @@ class MDnsVmServiceDiscovery {
       final Set<String> uniqueDomainNamesInResults = <String>{};
 
       // Listen for mDNS connections until timeout.
-      final Stream<PtrResourceRecord> ptrResourceStream = client.lookup<PtrResourceRecord>(
-        ResourceRecordQuery.serverPointer(dartVmServiceName),
-        timeout: timeout
-      );
+      final Stream<PtrResourceRecord> ptrResourceStream;
+
+      try {
+        ptrResourceStream = client.lookup<PtrResourceRecord>(
+          ResourceRecordQuery.serverPointer(dartVmServiceName),
+          timeout: timeout,
+        );
+      } on SocketException catch (e, stacktrace) {
+        _logger.printError(e.message);
+        _logger.printTrace(stacktrace.toString());
+        if (platform.isMacOS) {
+          throwToolExit(
+            'You may be having a permissions issue with your IDE. '
+                'Please try going to '
+                'System Settings -> Privacy & Security -> Local Network -> '
+                '[Find your IDE] -> Toggle ON, then restarting your phone.',
```

Comment:
Minor formatting nit:

```suggestion
            'You might be having a permissions issue with your IDE. '
            'Please try going to '
            'System Settings -> Privacy & Security -> Local Network -> '
            '[Find your IDE] -> Toggle ON, then restarting your phone.',
```